src/tap/src/tapcell.cpp

```diff
@@ -33,41 +33,1306 @@
+    tap_nw_masters.push_back(tap_nwin2_master);
+    tap_nw_masters.push_back(tap_nwin3_master);
+    tap_nw_masters.push_back(tap_nwouttie_master);
+    tap_nw_masters.push_back(tap_nwout2_master);
+    tap_nw_masters.push_back(tap_nwout3_master);
+
+    vector<string> tap_macro_masters;
+    tap_macro_masters.push_back(incnrcap_nwin_master);
+    tap_macro_masters.push_back(tap_nwin2_master);
+    tap_macro_masters.push_back(tap_nwin3_master);
+    tap_macro_masters.push_back(tap_nwintie_master);
+    tap_macro_masters.push_back(incnrcap_nwout_master);
+    tap_macro_masters.push_back(tap_nwout2_master);
+    tap_macro_masters.push_back(tap_nwout3_master);
+    tap_macro_masters.push_back(tap_nwouttie_master);
+
+    insertAtTopBottom(rows,
+                      tap_nw_masters,
+                      db_->findMaster(cnrcap_nwin_master),
+                      endcap_prefix_);
+    insertAroundMacros(rows,
+                       tap_macro_masters,
+                       db_->findMaster(cnrcap_nwin_master),
+                       endcap_prefix_);
+  }
+  insertTapcells(rows, tapcell_master, dist);
+  filled_sites_.clear();
+}
+
+void Tapcell::cutRows(odb::dbMaster* endcap_master,
+                      vector<odb::dbBox*> blockages,
+                      int& halo_x,
+                      int& halo_y)
+{
+  odb::dbBlock* block = db_->getChip()->getBlock();
+  int rows_count = (block->getRows()).size();
+  int block_count = blockages.size();
+  int min_row_width;
+  int end_width;
+
+  if (endcap_master != nullptr) {
+    end_width = endcap_master->getWidth();
+    min_row_width = 2 * end_width;
+  } else {
+    min_row_width = 0;
+  }
+
+  // Gather rows needing to be cut up front
+  vector<odb::dbRow*> blocked_rows;
+  std::map<string, vector<odb::dbBox*>> row_blockages;
+  string row_name;
+  for (odb::dbBox* blockage : blockages) {
+    for (odb::dbRow* row : block->getRows()) {
+      if (overlaps(blockage, row, halo_x, halo_y)) {
+        row_name = row->getName();
+        if (row_blockages.find(row_name) == row_blockages.end()) {
+          blocked_rows.push_back(row);
+          vector<odb::dbBox*> row_blockages_bbox;
+          row_blockages_bbox.push_back(blockage);
+          row_blockages.insert(std::pair<string, vector<odb::dbBox*>>(
+              row_name, row_blockages_bbox));
+        } else {
+          row_blockages[row_name].push_back(blockage);
+        }
+      }
+    }
+  }
+
+  // cut rows around macros
+  for (odb::dbRow* row : blocked_rows) {
+    cutRow(block, row, row_blockages, min_row_width, halo_x, halo_y);
+  }
+  int cut_rows_count = ((block->getRows()).size()) - rows_count;
+  logger_->info(utl::TAP, 1, "Found {} macro blocks.", block_count);
+  logger_->info(utl::TAP, 2, "Original rows: {}", rows_count);
+  logger_->info(utl::TAP, 3, "Created {} rows for a total of {} rows.", cut_rows_count, cut_rows_count+rows_count);
+}
+
+void Tapcell::cutRow(odb::dbBlock* block,
```

Comment:
Where is the current row being cut actually removed?

Reply:
At the end of the cutRow function the current row after it is cut is removed by odb::dbRow::destroy(row);